File: lib/golden_record/lanes.ex
defmodule GoldenRecord.Lanes do
  alias GoldenRecord.{Events, Cluster, IdentityLedger}

  @moduledoc """
  Typed entity lanes (gr-2a8): every code scheme belongs to exactly one entity type, identity
  claims route to their lane, and each lane folds its own ledger with a lane-qualified surrogate
  key prefix. Cross-lane bridging is structurally impossible — the lanes are disjoint folds, not
  a validation rule. `:uuid` is the one shared scheme (engine-minted identity for records born
  without a source code, see `Uuid`); an identity claim whose codes are all lane-neutral must
  carry an explicit `entity:` in its data.
  """

  @lanes [:product, :substance, :description, :media]

  # scheme => lane. Anything not listed is :product — every pre-lane scheme (cnk, gtin, isbn, …)
  # was a product code, so the default keeps existing logs and adapters meaning what they meant.
  @lane_of %{
    cas: :substance,
    unii: :substance,
    substance_id: :substance,
    text_id: :description,
    asset_id: :media,
    leaflet_id: :media
  }

  # Lane-qualified surrogate-key prefixes. :product keeps the legacy "SK" so existing logs,
  # fixtures, and customer-facing keys are unchanged by the lanes migration.
  @prefix %{product: "SK", substance: "SUB", description: "DSC", media: "MED"}

  @by_name Map.new(@lanes, &{Atom.to_string(&1), &1})

  def lanes, do: @lanes
  def prefix(lane), do: Map.fetch!(@prefix, lane)

  @doc ~s{Lane atom for a wire entity name ("description" => :description) — never an atom leak.}
  def parse(name), do: Map.fetch(@by_name, name)

  @doc """
  Lane of one code scheme. Runtime registry declarations win; built-in schemes remain the
  compatibility fallback. `:uuid` is shared (nil); undeclared schemes default to :product.
  """
  def lane_of_scheme(scheme), do: lane_of_scheme(scheme, nil)
  def lane_of_scheme(:uuid, _registry), do: nil
  def lane_of_scheme("uuid", _registry), do: nil

  def lane_of_scheme(scheme, %SchemeRegistry{} = registry),
    do: SchemeRegistry.lane(registry, scheme) || Map.get(@lane_of, scheme, :product)

  def lane_of_scheme(scheme, nil), do: Map.get(@lane_of, scheme, :product)

  @doc "Lane of a surrogate key, by its prefix (\"SUB_3\" => :substance)."
  def lane_of_key(key) do
    Enum.find(@lanes -- [:product], :product, &String.starts_with?(key, prefix(&1) <> "_"))
  end

  @doc """
  Lane of an identity claim: the unique lane among its codes' schemes (`:uuid` is neutral),
  falling back to an explicit `entity:` in the claim data, else :product. Codes from two lanes
  in one claim are a contract violation — `{:error, {:mixed_lanes, lanes}}`.
  """
  def of_claim(claim), do: of_claim(claim, nil)

  def of_claim(%Events.ClaimAsserted{kind: :identity, data: data}, registry) do
    data.codes
    |> Enum.map(fn {scheme, _} -> lane_of_scheme(scheme, registry) end)
    |> Enum.reject(&is_nil/1)
    |> Enum.uniq()
    |> case do
      [] -> {:ok, Map.get(data, :entity, :product)}
      [lane] -> {:ok, lane}
      lanes -> {:error, {:mixed_lanes, Enum.sort(lanes)}}
    end
  end

  @doc "The identity claims of one lane (mixed-lane claims belong to no lane)."
  def identity_claims(claims, lane), do: identity_claims(claims, lane, nil)

  def identity_claims(claims, lane, registry),
    do: Enum.filter(claims, &(&1.kind == :identity and of_claim(&1, registry) == {:ok, lane}))

  @doc false
  def identity_evidence(claims, lane, registry \\ nil) do
    Enum.filter(claims, fn
      %Events.ClaimAsserted{kind: :identity_evidence, data: %{left: {left, _}, right: {right, _}}} ->
        lane_of_scheme(left, registry) == lane and lane_of_scheme(right, registry) == lane

      _ ->
        false
    end)
  end

  @doc "Partition a ledger's members map by each key's lane."
  def partition_members(members) do
    grouped = Enum.group_by(members, fn {k, _codes} -> lane_of_key(k) end)
    Map.new(@lanes, fn lane -> {lane, Map.new(Map.get(grouped, lane, []))} end)
  end

  @doc "A fresh ledger per lane, each minting under its own prefix."
  def new_ledgers, do: Map.new(@lanes, &{&1, IdentityLedger.new(prefix(&1))})

  @doc """
  Cluster + reconcile each lane's identity claims against that lane's own ledger — the per-lane
  fold. Returns `{identity_events, ledgers}`; events come out in lane order (product first).
  """
  def reconcile(live_claims, shared, ledgers, at), do: reconcile(live_claims, shared, ledgers, at, nil)

  def reconcile(live_claims, shared, ledgers, at, registry),
    do: reconcile(live_claims, shared, ledgers, at, registry, %{})

  def reconcile(live_claims, shared, ledgers, at, registry, preferred) do
    Enum.flat_map_reduce(@lanes, ledgers, fn lane, acc ->
      case identity_claims(live_claims, lane, registry) do
        [] ->
          # A lane with NO identity claims is SKIPPED — members persist untouched. This fold
          # reads an append-only claim stream, possibly partial (a live batch may speak about
          # one lane only), so an empty lane means "no information", never "nothing exists".
          # History.reconcile_temporal deliberately does the OPPOSITE (retract): it replays a
          # complete effective snapshot, where absence IS evidence. Do not converge the two —
          # see gr-huw and test/reconcile_empty_lane_test.exs.
          {[], acc}

        claims ->
          clusters = Cluster.variants(claims ++ identity_evidence(live_claims, lane, registry), shared)
          lane_preferred = Map.take(preferred, clusters)
          events = IdentityLedger.decide(acc[lane], {:reconcile, clusters, shared, lane_preferred, at})
          {events, Map.put(acc, lane, Enum.reduce(events, acc[lane], &IdentityLedger.evolve(&2, &1)))}
      end
    end)
  end
end

File: php/src/Lanes.php
<?php

declare(strict_types=1);

namespace Ingot;

/**
 * Typed entity lanes — ported from `Lanes` in lib/golden_record_core.ex.
 *
 * Every code scheme belongs to exactly one entity type; identity claims route to their lane and
 * each lane folds its own ledger under a lane-qualified surrogate-key prefix. `uuid` is the one
 * shared (lane-neutral) scheme, so a claim made of only lane-neutral codes must carry an explicit
 * `entity`. Cross-lane bridging is structurally impossible — the lanes are disjoint folds.
 *
 * Prefer the {@see Lane} / {@see IdentityScheme} enums at call sites; the string helpers here remain
 * for the engine's internal wire shapes and for backward compatibility.
 */
final class Lanes
{
    /** @var list<string> */
    public const LANES = ['product', 'substance', 'description', 'media'];

    /** Non-product scheme => lane. Anything unlisted is 'product' (every pre-lane scheme was a product code). */
    private const LANE_OF = [
        'cas' => 'substance',
        'unii' => 'substance',
        'substance_id' => 'substance',
        'text_id' => 'description',
        'asset_id' => 'media',
        'leaflet_id' => 'media',
    ];

    /** @return list<string> */
    public static function lanes(): array
    {
        return array_map(static fn (Lane $lane): string => $lane->value, Lane::cases());
    }

    public static function prefix(string $lane): string
    {
        return Lane::from($lane)->prefix();
    }

    /** Lane atom for a wire entity name ("description" => 'description'), or null if unknown. */
    public static function parse(string $name): ?string
    {
        return self::tryParse($name)?->value;
    }

    public static function tryParse(string $name): ?Lane
    {
        return Lane::tryFrom($name);
    }

    /** Lane of one code scheme. 'uuid' is shared (null); unknown schemes default to 'product'. */
    public static function laneOfScheme(string $scheme): ?string
    {
        $lane = self::ofScheme($scheme);

        return $lane?->value;
    }

    /**
     * Typed lane of one code scheme. 'uuid' is shared (null); unknown schemes default to Product.
     * Synthetic lane-anchor schemes ({@see IdentityScheme}) resolve via their enum.
     */
    public static function ofScheme(string $scheme): ?Lane
    {
        if ($scheme === 'uuid') {
            return null;
        }

        $identity = IdentityScheme::tryFrom($scheme);
        if ($identity !== null) {
            return $identity->lane();
        }

        return isset(self::LANE_OF[$scheme])
            ? Lane::from(self::LANE_OF[$scheme])
            : Lane::Product;
    }

    /** Lane of a surrogate key, by its prefix ("SUB_3" => 'substance'). */
    public static function laneOfKey(string $key): string
    {
        return self::ofKey($key)->value;
    }

    public static function ofKey(string $key): Lane
    {
        foreach ([Lane::Substance, Lane::Description, Lane::Media] as $lane) {
            if (str_starts_with($key, $lane->prefix().'_')) {
                return $lane;
            }
        }

        return Lane::Product;
    }

    /**
     * Lane of an identity claim: the unique lane among its codes' schemes (uuid is neutral),
     * falling back to an explicit `entity` in the claim data, else 'product'. Two lanes in one
     * claim is a contract violation — returns ['error', ['mixed_lanes', sortedLanes]].
     *
     * @return array{0: string, 1: mixed}
     */
    public static function ofClaim(Claim $claim): array
    {
        $lanes = [];
        foreach ($claim->data['codes'] as $code) {
            $lane = self::laneOfScheme($code[0]);
            if ($lane !== null) {
                $lanes[$lane] = true;
            }
        }
        $lanes = array_keys($lanes);

        if ($lanes === []) {
            return ['ok', $claim->data['entity'] ?? 'product'];
        }
        if (count($lanes) === 1) {
            return ['ok', $lanes[0]];
        }
        sort($lanes);

        return ['error', ['mixed_lanes', $lanes]];
    }

    /**
     * The identity claims of one lane (mixed-lane claims belong to no lane).
     *
     * @param list<Claim> $claims
     * @return list<Claim>
     */
    public static function identityClaims(array $claims, string $lane): array
    {
        $out = [];
        foreach ($claims as $c) {
            if ($c->kind === 'identity' && self::ofClaim($c) === ['ok', $lane]) {
                $out[] = $c;
            }
        }

        return $out;
    }

    /**
     * Partition a ledger's members map by each key's lane. Returns lane => (key => code-set).
     *
     * @param array<string, array<string, array{0: string, 1: string}>> $members
     * @return array<string, array<string, array<string, array{0: string, 1: string}>>>
     */
    public static function partitionMembers(array $members): array
    {
        $out = array_fill_keys(self::LANES, []);
        foreach ($members as $key => $codes) {
            $out[self::laneOfKey($key)][$key] = $codes;
        }

        return $out;
    }

    /**
     * A fresh ledger per lane, each minting under its own prefix.
     *
     * @return array<string, LedgerState>
     */
    public static function newLedgers(): array
    {
        $out = [];
        foreach (self::LANES as $lane) {
            $out[$lane] = IdentityLedger::new(self::prefix($lane));
        }

        return $out;
    }

    /**
     * Cluster + reconcile each lane's identity claims against that lane's own ledger. Returns
     * [identityEvents, ledgers]; events come out in lane order (product first).
     *
     * @param list<Claim> $liveClaims
     * @param array<string, array{0: string, 1: string}> $shared a code-set
     * @param array<string, LedgerState> $ledgers
     * @return array{0: list<DomainEvent>, 1: array<string, LedgerState>}
     */
    public static function reconcile(array $liveClaims, array $shared, array $ledgers, mixed $at): array
    {
        $events = [];
        foreach (self::LANES as $lane) {
            $claims = self::identityClaims($liveClaims, $lane);
            if ($claims === []) {
                continue;
            }
            $evidence = array_values(array_filter(
                $liveClaims,
                static fn (Claim $claim): bool => $claim->kind === 'identity_evidence'
                    && self::laneOfScheme($claim->data['left'][0]) === $lane
                    && self::laneOfScheme($claim->data['right'][0]) === $lane,
            ));
            $clusters = Cluster::variants(array_merge($claims, $evidence), $shared);
            $laneEvents = IdentityLedger::decide($ledgers[$lane], ['reconcile', $clusters, $shared, $at]);
            $state = $ledgers[$lane];
            foreach ($laneEvents as $e) {
                $state = IdentityLedger::evolve($state, $e);
            }
            $ledgers[$lane] = $state;
            foreach ($laneEvents as $e) {
                $events[] = $e;
            }
        }

        return [$events, $ledgers];
    }
}
